Trace through `pytree` API with dynamo.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109433
* #109107
* __->__ #108533
* #109201

Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ysiraichi committed:
Trace through `pytree` API with dynamo.

Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

[ghstack-poisoned]

Skylion007 commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1294,6 +1312,11 @@ def call_id(self, tx, *args):
+            try:
+                value = args[0].as_python_constant()
+                return variables.ConstantVariable(id(value))
+            except:
+                pass
             unimplemented(f"call_id with args {args}")
```
Couldn't this unimplemented method be raised from the above exception so that we know why the args did not convert?

ezyang commented on torch/_dynamo/variables/misc.py:
```diff
@@ -933,3 +934,37 @@ def reconstruct(self, codegen):
     """Marker used to implement delattr()"""
+
+
+class TreeSpecVariable(VariableTracker):
```
Not sure why we need this variable tracker

ysiraichi replied:
You're right. I think we can make it a `ConstantVariable`.

zou3519 commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -21,7 +21,7 @@
 from ..guards import GuardBuilder
 from ..replay_record import DummyModule
-from ..source import AttrSource, is_constant_source, SuperSource, TypeSource
+from ..source import AttrSource, GetItemSource, is_constant_source, SuperSource, TypeSource
```
Since we are trying the new approach, it's probably easier if we undo the pieces of https://github.com/pytorch/pytorch/pull/108533 and go forward from there?

ysiraichi replied:
Not sure I get it. Undo the pieces of this PR?

zou3519 replied:
Yes. The approach taken in https://github.com/pytorch/pytorch/pull/108533 was to simulate pytrees in Dynamo by treating VariableTrackers as pytrees. We decided in https://github.com/pytorch/pytorch/issues/107315 that instead of the previous approach, we would like to move to one where we instead trace through all of the pytree code. So we are replacing the previous implementation with whatever the new implementation we're working on is.

ysiraichi replied:
Oh, wait. You mean #101896, right? #108533 is this current PR.
But, got it. I will remove those, leaving the tests (I believe they are still valid).

zou3519 replied:
Oh, sorry, yes that is what I meant

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Revert removal of pytree node registrations for VT. on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented on torch/utils/_pytree.py:
```diff
@@ -52,7 +51,7 @@ class NodeDef(NamedTuple):
     unflatten_fn: UnflattenFunc
 
-SUPPORTED_NODES: Dict[Type[Any], NodeDef] = {}
+SUPPORTED_NODES: Dict[int, NodeDef] = {}
```
meh but ok (but only because we typically don't expect types to get deallocated). I feel it shouldn't be that difficult to get Dynamo to support dict with Type input though...

ysiraichi replied:
The problem was how to materialize the sources for each type (+ `namedtuple` function) key in this dictionary. Would iterating through the type's `__qualname__` for that be a preferable solution?

ezyang replied:
This is for guards, right? I feel like we want something like https://github.com/pytorch/pytorch/pull/108528/ where we don't bother guarding on anything in SUPPORTED_NODES, and just observe when the dict is modified.

ezyang commented on torch/utils/_pytree.py:
```diff
@@ -336,9 +335,8 @@ def go(t):
-        @functools.wraps(f)
-        def inner(x: T) -> Any:
+    def deco(f: Any) -> Any:
+        def inner(x: Any) -> Any:
```
because the type annotations are not supported?

ysiraichi replied:
As far as I remember, in order to support type annotations, I had to implement `UserDefinedObjectVariable.as_python_constant`. That lead to some CI failures, of which I don't remember, right now. I will try to revert this and see what happens.

ysiraichi replied:
Nevermind. I just had to make `Any` a `TypingVariable`.

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/dicts.py:
```diff
@@ -489,14 +489,10 @@ def _dictvariable_unflatten(
+    ConstDictVariable, _dictvariable_flatten, _dictvariable_unflatten
+)
+fx_pytree.register_pytree_flatten_spec(
+    ConstDictVariable,
```
Can't we get rid of these now entirely?

ysiraichi replied:
I don't think so (at least, not as is). `FunctorchVmapHigherOrderVariable`, for example, uses pytree for (un)flattening the arguments (e.g. `ConstDictVariable`) into a list of `TensorVariable`.

zou3519 replied:
cc @kshitij12345, but I'm OK with breaking that code temporarily so that we do the right thing here.

ezyang commented on torch/onnx/_internal/fx/dynamo_graph_extractor.py:
```diff
@@ -45,7 +45,7 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_val, exc_tb):
         for class_type in self._extensions:
-            pytree.SUPPORTED_NODES.pop(class_type)
+            pytree.SUPPORTED_NODES.pop(id(class_type))
```
This suggests changing SUPPORTED_NODES is BC-breaking, you may need SUPPORTED_NODE_IDS

zou3519 commented on torch/_dynamo/variables/lists.py:
```diff
@@ -726,17 +726,10 @@ def _listvariable_unflatten(values: List[Any], context: pytree.Context) -> ListV
-        ListVariable,
-        _listvariable_flatten,
-    )
+pytree._register_pytree_node(
+    ListVariable, _listvariable_flatten, _listvariable_unflatten
+)
+fx_pytree.register_pytree_flatten_spec(ListVariable, _listvariable_flatten)
```
ListVariable, etc, shouldn't be pytree nodes. We made them pytree nodes in the previous design, but in the new design (where we are tracing through pytrees) these do not need to be pytree nodes anymore

ezyang commented on torch/_dynamo/variables/torch.py:
```diff
@@ -597,68 +597,20 @@ def fn_with_prim_types(x):
-            return torch.utils._pytree.tree_map(map_fn, tree)
+        elif (
+            hasattr(self.value, "__module__")
+            and self.value.__module__ == "torch.utils._pytree"
```
It doesn't seem to me that you should need the case here. Since you just want to inline it, you want to arrange for these functions in pytree to have been wrapped as user function to begin with, and the call function would hit it automatically. This is a skipfiles thing. @yanboliang can give more advice.

ysiraichi commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1084,6 +1091,20 @@ def call_getattr(
 
+        if isinstance(obj, variables.ConstantVariable) and is_treespec_cls(
+            obj.python_type()
+        ):
```
We need to check whether this is a `ConstantVariable(TreeSpec)` and handle its `getattr` calls. Otherwise we'll get `GetAttrVariable(spec, name)`:
- `type`: can't call `id` on it
- `children_specs`: can't call `iter` on it

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi commented:
Here is a summary of the most recent changes:

- Add `expectedFailure` to `test_vmap_pytree_inputs` test
- Add `torch.utils._pytree` into the `disallowed_modules` tuple
- Special case `VariableBuilder` for dictionaries with type (+`namedtuple` function) as keys
    - Instead of a `DICT_KEYS`, we issue the new `DICT_VERSION` guard
- Introduce `ignore_guards` parameter for `VariableBuilder`
    - Used when we added `DICT_VERSION` guard, which assumes the values remain unchanged
    - This is controlled by `keys_are_types_or_namedtuple`, which is very specific
        - Maybe we should change it for `value is SUPPORTED_NODES`?
- Remove old `VariableTracker` registration for pytree API
- Turn `NodeDef` and `TreeSpec` instances into `UserDefinedObjectVariable`
    - Allows us to delete a bunch of `getattr` specializations for them

 @ezyang @zou3519 
Let me know what you think.

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented:
ci failures look legit

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi commented:
Right. I think I took care of the `TypeVar` issue. But, I'm not sure where the `test_layout_SparseCSR` failure is coming from. I will investigate it.

ezyang commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -2582,6 +2582,7 @@ def fn(x, y):
         )
         self.assertEqual(actual, expected)
 
+    @unittest.expectedFailure
```
Are you going to fix this later or what?

ezyang commented on torch/utils/_pytree.py:
```diff
@@ -337,7 +336,6 @@ def go(t):
     You can also directly use 'tree_map_only'
     """
     def deco(f: Callable[[T], Any]) -> Callable[[Any], Any]:
-        @functools.wraps(f)
```
meh, ok. Ideally dynamo can deal with this though

zou3519 replied:
@kshitij12345 has an upcoming fix for this

ezyang commented on torch/_dynamo/variables/constant.py:
```diff
@@ -81,6 +81,12 @@ def unpack_var_sequence(self, tx):
+                UserErrorType.ANTI_PATTERN,
+                "Can't access members of type(obj) for a generated custom object. "
+                "Please use __class__ instead",
+            )
```
I still don't really understand this error. Is this a notimplemented in disguise?

ysiraichi replied:
I think so. I just pushed it further, as suggested [in this comment](https://github.com/pytorch/pytorch/pull/98366/files#r1159296709).

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -406,6 +416,10 @@ def _wrap(self, value):
+            elif keys_are_types_or_namedtuple:
+                # For keys that are types, we only guard on the wheather the dictionary
+                # was modified by checking its version.
+                guards = self.make_guards(GuardBuilder.DICT_VERSION)
```
You need an id test too then, right?

ezyang replied:
(In fact, DICT_VERSION should probably imply the id test.)

ysiraichi replied:
Agreed.

ysiraichi commented:
I'm not really sure why [`test_layout_Sparse*` is erroring](https://github.com/pytorch/pytorch/actions/runs/6123423797/job/16621747198). I will try rebasing.

ysiraichi committed:
Rebased. on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi commented on benchmarks/dynamo/ci_expected_accuracy/inductor_torchbench_inference.csv:
```diff
@@ -10,7 +10,7 @@ basic_gnn_gcn,pass,6
 basic_gnn_sage,pass,0
 clip,pass,0
-cm3leon_generate,pass,8
+cm3leon_generate,pass,7
```
Due to the handling `typing.Optional` as `TypingVariable`.

ezyang replied:
I'll never turn down a reduction in graph breaks ;)

voznesenskym replied:
separate PR?

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -205,6 +205,7 @@ def __init__(
         self,
         tx,
         source: Source,
+        ignore_guards: bool = False,
```
Doc this? When is it OK to set this to true?

ezyang replied:
Make this kwarg only

ezyang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1101,6 +1107,21 @@ def call_getattr(
                 unimplemented("tensor grad")
+        elif name == "__bases__":
+            value = obj.as_python_constant()
+            if isinstance(value, type):
```
You can't do it this way, this will cause the rest of the elif sled to get bypassed

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -393,9 +398,14 @@ def _wrap(self, value):
             ConstantVariable.is_literal(k)
             or self.tensor_can_be_dict_key(k)
-            or isinstance(k, enum.Enum)
+            or isinstance(k, (enum.Enum, type))
```
this seriously need a refactor I think... (no action needed here)

ezyang reviewed (approved):
please fix the elif thing

voznesenskym commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1101,6 +1107,21 @@ def call_getattr(
                 unimplemented("tensor grad")
             else:
                 unimplemented("tensor grad")
+        elif name == "__bases__":
```
I dont understand why `__bases__` access is just being handled as a call_getattr to a builtin like this - can you say more?

ysiraichi replied:
Because it's the first thing that is called on `LOAD_ATTR`. Otherwise, I would have to copy-paste the code for the VTs that could represent a type (e.g. `ConstantVariable`, `UserDefinedClassVariable`).

voznesenskym replied:
This is fine

voznesenskym commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1101,6 +1107,21 @@ def call_getattr(
+            if isinstance(value, type):
+                bases = value.__bases__
+                if source is None:
+                    assert len(bases) == 1 and (
```
I don't understand this assert - when would the base of a builtin be a TensorBase?

ysiraichi replied:
**Why do we need it?** If we don't have a source (i.e. can't guard on it), the result of `<type>.__bases__` should be one of those.

**When would the base of a builtin be a `TensorBase`?**
- _`TestExport.test_export_preserve_signature` @ test/export/test_export.py_ needs it
- Whenever, for some reason, we use pytree on sourceless tensors

ysiraichi replied:
@voznesenskym @ezyang 
Question: if we don't have a source, how to make sure that `bases` is still either `(object,)` or `(TensorBase,)`?

voznesenskym replied:
If we don't have a source, you do not need to guard on the type. Its an intermediary, meaning it is a product of some other operations, which should be protected via guard on input, state, etc.

voznesenskym commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1101,6 +1107,21 @@ def call_getattr(
+            value = obj.as_python_constant()
+            if isinstance(value, type):
+                bases = value.__bases__
+                if source is None:
```
If you don't have a source, and you want a VT you need to use SourcelessBuilder

voznesenskym commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1101,6 +1107,21 @@ def call_getattr(
+                    tuple_args = [BuiltinVariable(bases[0])]
+                else:
+                    tuple_args = [
+                        VariableBuilder(tx, GetItemSource(source, i))(b)
```
Your source here is skipping the `__bases__` access, which, I am not sure is what you want?

ysiraichi replied:
The `AttrSource` is being created before is if-else chain.

voznesenskym commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -518,6 +518,18 @@ static PyObject* check_obj_id(PyObject* dummy, PyObject* args) {
   }
 }
 
+static PyObject* dict_version(PyObject* dummy, PyObject* args) {
```
nice

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -419,9 +437,16 @@ def index_source(key):
 
+            # If keys_are_types_or_namedtuple is True, don't guard on dictionary values.
+            #
+            # In this case, we are guarding on the dictionary version (i.e. it wasn't
```
While you can ignore guards here because you are guarding elsewhere - I strongly dislike the pattern that `ignore_guards` creates. It's a huge hammer to just be able to pass like this, imo. I would rather we build `result` differently, or build it and then strip the guards off.

ysiraichi replied:
I agree. So, in order not to leave `ignore_guards`, I will specialize a case for `SUPPORTED_NODES`.

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi commented:
@voznesenskym Let me know what you think.

ezyang commented:
@pytorchbot merge -i

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi commented:
@pytorchbot merge

ysiraichi commented:
@pytorchbot merge

ysiraichi commented:
@pytorchbot merge -i

ysiraichi committed:
Rebased. on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -r main

ysiraichi committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zhxchen17 commented:
@pytorchbot merge -i

ezyang committed:
Update on "Trace through `pytree` API with dynamo."


Fix: #107315

This PR enables dynamo to trace through the `pytree` API by inlining its functions. In
order to do so, a few details of `pytree` had to be changed.

In summary, this PR:

- Introduces `TreeSpecVariable` for representing `TreeSpec` instances
- Specializes `<type>.__bases__` call, returning a `TupleVariable`
- Enables the call to `id` builtin function for every variable that implements
  `as_python_constant` method
- Specializes `ConstantVariable.call_method` for its (un)flatten functions
- Implements `UserDefinedObjectVariable.as_python_constant`
- Modifies `pytree` by:
    - Make `SUPPORTED_NODES` a map of ids (instead of types) to `NodeDef`
    - Removed `functools.wraps` function, since it can't be inlined

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
